### PR TITLE
Removed unneeded `.detach` in `clip_grad_norm_`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120617
* #120238
* __->__ #120613
* #120612
* #120231

This PR removes some unneeded `.detach()` calls in `clip_grad_norm_()`.

In the future, we prefer to remove the `g.detach().mul_()` and use `torch.no_grad()` instead, but that is blocked by a circular import issue.

### CODE REVIEW

**Comment by awgu** [comment]
Closing this to let @janeyx99 land the proper approach
